File: fixtures/scale/dart/large/lib/ciphers/columnar/columnar_cipher.dart
import '../../core/cipher.dart';
import '../../util/alphabet.dart';
import 'columnar_key.dart';

class ColumnarCipher implements Cipher {
  final ColumnarKey key;
  final List<int> _order;

  ColumnarCipher(this.key) : _order = key.columnOrder();

  @override
  String name() => 'columnar';

  /// Write in rows, read the columns in keyword order.
  @override
  String encrypt(String plaintext) {
    final text = key.padded(cleanAlphabet(plaintext));
    final sb = StringBuffer();
    for (final col in _order) {
      var row = 0;
      while (row * _order.length + col < text.length) {
        sb.write(text[row * _order.length + col]);
        row++;
      }
    }
    return sb.toString();
  }

  @override
  String decrypt(String ciphertext) {
    final text = cleanAlphabet(ciphertext);
    final rows = text.length ~/ _order.length;
    final out = List<String>.filled(text.length, '');
    var cursor = 0;
    for (final col in _order) {
      for (var row = 0; row < rows; row++) {
        out[row * _order.length + col] = text[cursor++];
      }
    }
    return out.join();
  }
}

File: fixtures/scale/kotlin/large/src/cryptobench/ciphers/columnar/ColumnarCipher.kt
package cryptobench.ciphers.columnar

import cryptobench.core.Cipher
import cryptobench.util.Alphabet

class ColumnarCipher(private val key: ColumnarKey) : Cipher {
    private val order = key.columnOrder()

    override fun name(): String = "columnar"

    /** Write in rows, read the columns in keyword order. */
    override fun encrypt(plaintext: String): String {
        val text = key.padded(Alphabet.clean(plaintext))
        val sb = StringBuilder()
        for (col in order) {
            var row = 0
            while (row * order.size + col < text.length) {
                sb.append(text[row * order.size + col])
                row++
            }
        }
        return sb.toString()
    }

    override fun decrypt(ciphertext: String): String {
        val text = Alphabet.clean(ciphertext)
        val rows = text.length / order.size
        val out = CharArray(text.length)
        var cursor = 0
        for (col in order) {
            for (row in 0 until rows) {
                out[row * order.size + col] = text[cursor++]
            }
        }
        return String(out)
    }
}
